File: experiments/cbb-emergency-access/reference/src/tunnel/forwarder.mjs
/**
 * 默认转发器适配器：**演练实现**。
 *
 * 它只维护规则表，不建立真实 SSH 连接——因此禁止用于生产（见 constraints.md 的部署约束）。
 * 生产部署必须由 配置：tunnel.forwarder_module 指向真实转发实现（例如以 ssh -L 建立端口转发），
 * 该实现必须满足 tunnel/spec.md 里同一个适配器契约，从而复用同一批判据。
 */
const rules = new Map();
let sequence = 0;

export function createForwarder() {
  return {
    open({ session_id, server_id, source_ip }) {
      sequence += 1;
      const rule_id = 'rule-' + sequence;
      rules.set(rule_id, { rule_id, session_id, server_id, source_ip, opened_at: new Date().toISOString() });
      return { rule_id };
    },
    close({ session_id }) {
      for (const [rule_id, rule] of rules) {
        if (rule.session_id === session_id) rules.delete(rule_id);
      }
    },
    list() {
      return [...rules.keys()];
    },
  };
}

File: experiments/cbb-emergency-access-go/internal/tunnel/forwarder.go
// 内置的转发器适配器：只维护规则表、不建立真实 SSH 连接。
//
// 这是演练用的实现（specs/constraints.md 允许的默认适配器）；生产部署要用真实转发实现。
// 规则表刻意放在进程内存里并跨会话共享——真实世界里规则活在 SSH 主机上，
// 本服务重启不会让它们消失，这一点在 A14（跨进程重启后回收）里被验到。
//
// 它是接缝的一部分（specs/config.md「接缝」表里的「转发器」一行），所以属于种子，
// 不属于任何 capability 任务。
package tunnel

import (
	"fmt"
	"sync"

	"cbb-emergency-access/internal/contract"
)

type memoryForwarder struct {
	mu    sync.Mutex
	seq   int
	rules map[string]contract.Rule
}

// NewMemoryForwarder 返回演练用的内存转发器。
func NewMemoryForwarder() contract.Forwarder {
	return &memoryForwarder{rules: map[string]contract.Rule{}}
}

func (f *memoryForwarder) Open(spec contract.ForwardSpec) (contract.Rule, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.seq++
	rule := contract.Rule{
		ID:         fmt.Sprintf("rule-%d", f.seq),
		SessionID:  spec.SessionID,
		ServerID:   spec.ServerID,
		SourceIP:   spec.SourceIP,
		ListenPort: spec.ListenPort,
	}
	f.rules[rule.ID] = rule
	return rule, nil
}

func (f *memoryForwarder) Close(sessionID string) error {
	f.mu.Lock()
	defer f.mu.Unlock()
	for id, rule := range f.rules {
		if rule.SessionID == sessionID {
			delete(f.rules, id)
		}
	}
	return nil
}

func (f *memoryForwarder) List() ([]contract.Rule, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	out := make([]contract.Rule, 0, len(f.rules))
	for _, rule := range f.rules {
		out = append(out, rule)
	}
	return out, nil
}
